File: app/javascript/components/mutations/registerMutation.js
import gql from 'graphql-tag';

export default gql`
  mutation register($username: String!, $password: String!, $passwordConfirmation: String!) {
    registerMutation(username: $username, password: $password, passwordConfirmation: $passwordConfirmation) {
      user {
        username
      }
    }
  }
`;

File: app/graphql/mutations/register_mutation.rb
class Mutations::RegisterMutation < GraphQL::Schema::Mutation
  description 'This is only a test mutation, should be removed'

  argument :username, String, required: true
  argument :password, String, required: true
  argument :passwordConfirmation, String, required: true

  field :user, Types::UserType, null: false

  def resolve(args)
    if !context[:current_user] || context[:current_user].admin?
      user = User.create(args)
      if user.valid?
        { user: user }  
      else
        raise GraphQL::ExecutionError, user.errors.full_messages.join(', ')
      end
    else
      raise GraphQL::ExecutionError, 'Not allowed'
    end
  end
end
